pkg/kubeutils/kubeutils.go

```diff
@@ -117,44 +117,70 @@ func (k *KubeUtils) WaitForJob(ctx context.Context, cli client.Client, ns, name
 	var lasterr error
 	if err := wait.ExponentialBackoffWithContext(
 		ctx, backoff, func(ctx context.Context) (bool, error) {
-			ready, err := k.IsJobComplete(ctx, cli, ns, name, completions)
-			if err != nil {
-				lasterr = fmt.Errorf("unable to get job status: %w", err)
+			var job batchv1.Job
+			err := cli.Get(ctx, client.ObjectKey{Namespace: ns, Name: name}, &job)
+			if k8serrors.IsNotFound(err) {
+				// exit
+				lasterr = fmt.Errorf("job not found")
+				return false, lasterr
+			} else if err != nil {
+				lasterr = fmt.Errorf("unable to get job: %w", err)
 				return false, nil
 			}
-			return ready, nil
+
+			failed := k.isJobFailed(job)
+			if failed {
+				// exit
+				lasterr = fmt.Errorf("job failed")
+				return false, lasterr
+			}
+
+			completed := k.isJobCompleted(job, completions)
+			if completed {
+				return true, nil
+			}
+
+			// TODO: need to handle the case where the pod get stuck in pending
+			// This can happen if nodes are not schedulable or if a volume is not found
+
+			return false, nil
 		},
 	); err != nil {
 		if lasterr != nil {
-			return fmt.Errorf("timed out waiting for job %s: %w", name, lasterr)
-		} else {
-			return fmt.Errorf("timed out waiting for job %s", name)
+			return lasterr
 		}
+		return fmt.Errorf("timed out waiting for job %s", name)
 	}
 	return nil
 }
 
 // WaitForPod waits for a pod to be completed.
-func (k *KubeUtils) WaitForPodComplete(ctx context.Context, cli client.Client, ns, name string, opts *WaitOptions) error {
+func (k *KubeUtils) WaitForPodComplete(ctx context.Context, cli client.Client, ns, name string, opts *WaitOptions) (*corev1.Pod, error) {
```

Comment:
my first impression when reading this function name is that it would return an error if the pod fails, but looks like it's up to the caller to make that decision based on the returned pod status?

Reply:
Completed != succeeded